docs: add notice about `address` conversion prior to 0.8.0

CJ42 opened the pull request:
# What does this PR introduce?

Prior to Solidity 0.8.0, it was possible to **convert explicitly** any integer type (signed or unsigned) to the `address` type.

For instance, these Solidity code snippets would have compiled up to version 0.7.6:

```solidity
pragma solidity 0.7.6;

contract Address {

    function convertAddress(uint256 a) public pure returns (address) {
        return address(a);
    }

    function convertAddress(uint48 a) public pure returns (address) {
        return address(a);
    }

    function convertAddress(int8 a) public pure returns (address) {
        return address(a);
    }
    
}
```

## Docs changed introduced

This PR aims to clarify the change of behaviour in the Section **Types > Conversion > Address**, specifying that:

- explicit conversion from `uint160` is allowed.
- add a notice about the behaviour of explicit `address` conversion prior to 0.8.0.

Your commit messages should detail why you made your change in addition to what you did (unless it is a tiny change).

CJ42 committed:
docs: add `uint160` type in address conversion section

docs: add notice for uint conversion before 0.8.0

docs: edit sentence structure + remove 'result in `address payable`'

docs: add keyword 'explicitly' to be more specific for `address payable`

docs: edit notice uint to address conversion change since 0.8.0

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

cameel commented on docs/types/conversion.rst:
```diff
@@ -188,6 +188,11 @@ Addresses
 test are of ``address`` type. No other literals can be implicitly converted to the ``address`` type.
 
-Explicit conversions from ``bytes20`` or any integer type to ``address`` result in ``address payable``.
+Explicit conversions from ``bytes20`` or ``uint160`` to ``address`` is allowed and result in ``address payable``.
```
```suggestion
Explicit conversions to ``address`` are allowed only from ``bytes20`` and ``uint160`` and result in ``address payable``.
```

cameel commented on docs/types/conversion.rst:
```diff
@@ -188,6 +188,11 @@ Addresses
 test are of ``address`` type. No other literals can be implicitly converted to the ``address`` type.
 
-Explicit conversions from ``bytes20`` or any integer type to ``address`` result in ``address payable``.
+Explicit conversions from ``bytes20`` or ``uint160`` to ``address`` is allowed and result in ``address payable``.
```
This part is actually outdated. Starting with 0.8.0 you always need an explicit conversion to get a payable address.
```suggestion
Explicit conversions from ``bytes20`` or ``uint160`` to ``address`` are allowed.
```

CJ42 replied:
@cameel 
I have addressed your comment and removed this last part from the sentence.

I have suggested to mention it in the next sentence then, to be more specific. See this commit: https://github.com/ethereum/solidity/pull/13105/commits/5f2be0c8b776d9eff750fd6bcb023a7e949e6c10

Let me know if that's ok, otherwise I will remove it.

cameel commented on docs/types/conversion.rst:
```diff
@@ -188,6 +188,11 @@ Addresses
+.. note::
+    Prior to version 0.8.0, it was possible to convert explicitly any integer type (signed or unsigned) of any size 
+    (from ``int8``/``uint8`` to ``int256``/``uint256``) to an address. This was removed in 0.8.0, and only conversion 
+    from ``uint160`` is now allowed.
```
```suggestion
    Prior to version 0.8.0, it was possible to explicitly convert from any integer type (of any size, signed or unsigned) to  ``address`` or ``address payable``.
    Starting with in 0.8.0 only conversion from ``uint160`` is allowed.
```

cameel commented on docs/types/conversion.rst:
```diff
@@ -188,6 +188,11 @@ Addresses
 An ``address a`` can be converted to ``address payable`` via ``payable(a)``.
+
+.. note::
+    Prior to version 0.8.0, it was possible to convert explicitly any integer type (signed or unsigned) of any size 
+    (from ``int8``/``uint8`` to ``int256``/``uint256``) to an address. This was removed in 0.8.0, and only conversion 
```
`chk_coding_style` job is failing because of this:
```suggestion
    Prior to version 0.8.0, it was possible to convert explicitly any integer type (signed or unsigned) of any size
    (from ``int8``/``uint8`` to ``int256``/``uint256``) to an address. This was removed in 0.8.0, and only conversion
```
Please configure your editor to automatically remove trailing whitespace.

CJ42 commented:
@cameel
I have addressed your comment and removed this last part from the sentence (about _"result in `address payable`")_

I have suggested to mention it in the next sentence then, to be more specific. See this commit: https://github.com/ethereum/solidity/commit/5f2be0c8b776d9eff750fd6bcb023a7e949e6c10

Let me know if that's ok, otherwise I will remove it.

cameel commented:
Looks good. Can you squash your commits so that we can merge this?

CJ42 commented:
@cameel I have squashed my commits as requested :)

CJ42 commented:
@cameel some tests in the CI are failing. I am not sure to understand why.

cameel commented:
Is it actually rebased on top of `develop`?

BTW, I'm off currently but maybe someone else from the team could help with the rebase here? @matheusaaguiar?

matheusaaguiar commented:
Yes, it needs a rebase.
@CJ42 , you can do it with the following commands:
```
git remote add upstream https://github.com/ethereum/solidity.git  #add solidity remote repo
git fetch upstream                                                #update from solidity remote repo
git rebase upstream/develop                                #rebase on top of current develop
git push origin develop -f                                      #force update on your fork repo 
```
Let me know if need any help.

matheusaaguiar commented:
[This](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) should help too if you more used to the web UI.

CJ42 commented:
@matheusaaguiar I think I messed up the rebase. Fetched and merged from the Github UI in my fork, and now the commit history is not clean anymore.
Is there a place to chat for support on Discord or Gitter?

matheusaaguiar commented:
@CJ42 Nice! Don't worry about these failing tests, they are broken due to reasons unrelated to your PR.

matheusaaguiar reviewed (dismissed):
Looks good.

CJ42 commented:
@matheusaaguiar @cameel is there anything else to do on my side so that this can be merged?

matheusaaguiar commented:
Sorry about the delay, @CJ42 . I rebased your PR and it should be merged once all tests are green.
